applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_POSITION UCHAR_MAX
+#define DEBOUNCE_DELAY K_MSEC(5)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+	STATE_ERROR,
+};
+
+struct selector {
+	const struct selector_config *config;
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t position;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(selector_config)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	__ASSERT_NO_MSG((selector->position < selector->config->pins_size) ||
+			(selector->position == DEFAULT_POSITION));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_id = selector->config->id;
+	event->position = selector->position;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value = 0;
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int err = 0;
```

Comment:
`int err;`

Reply:
Ok, also added assert to ensure that numer of selectors is non-zero - otherwise we may have undefined behaviour here